engine/main.ml

```diff
@@ -224,39 +230,337 @@ let rec parse_lambda lsp =
+  let rec bprint_tree ntabs buf tree =
+    let sep = break ntabs in
+    match tree with
+    | Leaf (pilist, target_blackbox) ->
+       bprintf buf
+         "Leaf=%s\
+              %t%a"
+         target_blackbox
+         (break ntabs) (bprint_list ~sep bprint_pi) pilist
+    | Node (cst_list) ->
+       bprintf buf
+         "Node=\
+          %t%a"
+         (break ntabs)
+         (bprint_list ~sep (bprint_tree (ntabs+1))) cst_list
+  and
+  bprint_env ntabs buf env =
+    let bprint_binding buf (key, entry) =
+      let bprint_value buf = function
+        | Accessor a -> bprint_accessor buf a
+        | Addition (v, i) ->
+           bprintf buf "Addition=%s,%i" v i
+        | Function (v, f_tree) ->
+           bprintf buf  "Function=%s,ConstraintTree: %a"
+             v
+             (bprint_tree (ntabs + 1)) f_tree
+        | Catch (e, vars, c_tree) ->
+          bprintf buf "Catch=%d %s,ConstraintTree: %a"
+            e
+            (String.concat " " vars)
+             (bprint_tree (ntabs + 1)) c_tree
+      in
+      bprintf buf "%s: %a" key bprint_value entry
+    in
+    bprint_list ~sep:(break ntabs) bprint_binding buf (SMap.bindings env)
+  in
+  let print_env env =
+    let buf = Buffer.create 42 in
+    bprint_env 0 buf env;
+    BatIO.write_line BatIO.stdout (Buffer.contents buf)
+  in
+  let expand_env variable entry : environment SMap.t = 
+    let _ = match SMap.find_opt variable env with
+      | Some _ -> assert false
+      | None -> ()
+    in
+    SMap.add variable entry env
+  in
+  (* perform union on two maps, keys should never differ *)
+  let union env1 env2 = SMap.union (fun _ a b -> assert (a = b); Some a) env1 env2
+  in
+  let match_let_accessor env acc =
+    match acc with
+    | Var v -> Accessor (AcRoot v)
+    | Field (i, v) ->
+      let acc = match SMap.find_opt v env with
+        | Some (Accessor a) -> a
+        | Some (Addition _) -> assert false
+        | Some (Function _) -> assert false
+        | Some (Catch _) -> assert false
+        | None -> assert false
+      in
+      Accessor (AcField (acc, i))
+    | Function (v, sxp) ->
+      let env' = expand_env v (Accessor (AcRoot v))
+      in
+      let constraint_tree = sym_exec sxp constraints env' in
+      Function (v, constraint_tree)
+    | _ -> assert false
+  in
+  match sexpr with
+  | Let (blist, next_sexpr) ->
+    let env' = blist |>
+               List.map (fun (var, sxp) -> expand_env var (match_let_accessor env sxp)) |>
+               List.fold_left union env
+    in
+    sym_exec next_sexpr constraints env'
+  | If (bexpr, strue, sfalse) ->
+    let piop, sxp =
+      match bexpr with
+      | Comparison (bop, sxp, i) -> (match_bop (bop, i)), sxp
+      | Isout (i, v) -> Isout i, Var v
+      | _ -> assert false (* TODO *)
+    in
+    let avar = match sxp with
+      | Var v -> begin
+          match SMap.find_opt v env with
+        | Some (Accessor a) -> a
+        | _ -> assert false
+        end
```

Comment:
so `SMap.find` would be shorter here (than using `find_opt` and then `assert false`)